[Ruby] New Concept exercise: numbers (#1295)
exercism/v3: languages/ruby/exercises/concept/numbers/.docs/introduction.md

There are two different types of numbers in Ruby:

- Integers: numbers with no digits behind the decimal separator (whole numbers). Examples are `-6`, `0`, `1`, `25`, `976` and `500000`.
- Floating-point numbers: numbers with zero or more digits behind the decimal separator. Examples are `-2.4`, `0.1`, `3.14`, `16.984025` and `1024.0`.

They are implemented through the [`Integer`][integer-ruby] and [`Float`][float-ruby] class.
Arithmetic is done using the standard arithmetic operators. Numbers can be compared using the standard numeric comparison operators(`>`, `<=`, `==`, `!=`, etc.).
Basic arithmetic operations between instances of `Integer`, will always result in an instance of `Integer`:

> ###`+`

Reviewer: One thing to note is that students that will be working on this exercise, must have completed the `basics` exercise (speaking of which, you should update Ruby's [`config.json` file](https://github.com/exercism/v3/blob/master/languages/ruby/config.json)). As such, they'll be familiar with some basic arithmetic operators like `*` and `+` (see the [basics example](https://github.com/exercism/v3/blob/master/languages/ruby/exercises/concept/basics/.meta/example.rb) and the [basics after.md](https://github.com/exercism/v3/blob/master/languages/ruby/exercises/concept/basics/.docs/after.md)). Could you take a look at which things you can assume a student will already know and remove what the student already knows?

Another thing to note is that the `introduction.md` file should contain just the minimal amount of information for a student to solve it. Maybe you could take a look at whether everything in the introduction is required for the student to know in order to complete the exercise? If it is not required, it should be removed from the introduction. Note that the `after.md` allows for a more expansive description of the concept, and can also contain non-essential knowledge.

```ruby
a = 1
b = 2.0
a.class
#=> Integer
b.class
#=> Float
c = a + b
#=> 3.0 (Notice the decimal separator)
c.class
#=> Float
d = b / 2
#=> 1.0
d.class
#=> Float
```

Both instances of `Float` and `Integer` implement methods to convert values from one to the other. As a number implemented through the `Integer` class has less precision than a number implemented through the `Float` class, converting from an instance of `Integer` to an instance of `Float` is safe because it does not result in any kind of data loss. However, converting from an instance of `Float` to an instance of `Integer` could mean losing data.

In this exercise you must conditionally execute logic. A common way to do this in Ruby is by using an `if/else` statement:

```ruby
x = 5

if x == 5
  # Execute logic if x equals 5
elsif x > 7
  # Execute logic if x greater than 7
else
  # Execute logic in all other cases
end
```

The condition of an `if` statement does not have to be only `true` or `false`. It can be any value. But it's important to know that any value other than `nil` and `false` (_truthy_ values) will be treated as `true`, meaning that the code inside the `if` statement will be executed.

> Ruby also implements the `case` statement:

Reviewer: Having this description of the `case` statement implies to me that we expect the students to use a `case` statement in the exercise. I don't see if used in the example file though, so I'm wondering if this is the right place to introduce the `case` statement (if at all).

Author: Good point. Since the exercise is covering the `conditionals` concept, is it ok to mention the `case` statement in the `after.md` then? I feel like it would be bad to not mention it at all or have a separate exercise just for the `case` statement (or even just `conditionals`).

Reviewer 2: I think mentioning it in after.md is OK.  I don't think it should be suggested that it should have been used though.

Reviewer: I'm fine with including it in the `after.md` document.

> There are two different types of numbers in Ruby:

Reviewer: > There are two different types of numbers in Ruby, just like there are two different types of ducks. But really there are more than two different types of ducks, just like there are more than two different types of numbers in Ruby.

I don't completely understand this comment. Are there numbers that are not integers or floating-point numbers in Ruby?

> There are two different types of numbers in Ruby:

Reviewer 2: Yes.  which is why I state we are only looking at two of them.  Line 5 of `design.md` does state "two of the" so perhaps this is covered there.

I was concerned that perhaps we were implying that there are only two kinds of numbers in Ruby.  But as noted in line 5 of `design.md` that may very well be enough to be noted.

Just for reference here:

```ruby
>> numeric_types = [1, 1.0, 1r, 1+0i]
=> [1, 1.0, (1/1), (1+0i)]
>> numeric_types.each do |number|
..     puts '%s is type %s and is Numeric: %s' % [number, number.class, number.is_a?(Numeric)]
..   end
1 is type Integer and is Numeric: true
1.0 is type Float and is Numeric: true
1/1 is type Rational and is Numeric: true
1+0i is type Complex and is Numeric: true
=> [1, 1.0, (1/1), (1+0i)]
```

> There are two different types of numbers in Ruby:

Reviewer: Ruby has built in rational and complex numbers? That's awesome! Thanks for explaining.

> There are two different types of numbers in Ruby:

Reviewer 3: ```suggestion
The two mainly used types of numbers in Ruby are:
```

> There are two different types of numbers in Ruby:

Reviewer 2: ```suggestion
Two common types of numbers in Ruby:
```

Reviewer 2: Mostly because, though I intuitively agree that it may be possible these are the two most common, I have not measured that to find out.

Reviewer: > I have not measured that to find out

It seems like a real stretch to think of any scenario where rational or complex numbers are more common than integers and floats, but I also don't have any data to back that up.

Reviewer 2: I could think of no scenario where float and integer wouldn't outweigh the "rational **or** complex", but I can think of some where it may be comparable or perhaps exceeded even, if we use **and** instead.

- Integers: numbers with no digits behind the decimal separator (whole numbers). Examples are `-6`, `0`, `1`, `25`, `976` and `500000`.
- Floating-point numbers: numbers with zero or more digits behind the decimal separator. Examples are `-2.4`, `0.1`, `3.14`, `16.984025` and `1024.0`.

> They are implemented through the `Integer` and `Float` class.

Reviewer 2: We might mention the parent class `Numeric` here.  The concept is "Numbers" so we should explain where they come from, that some of the behavior they have come from the parent class.

Reviewer: > We might mention the parent class Numeric here. The concept is "Numbers" so we should explain where they come from, that some of the behavior they have come from the parent class.

I would argue against mentioning that. Knowing that these classes have a shared base means knowing what inheritance inheritance is and how it works, which we don't know for certain that students will be familiar with at this point. I think this information should either be mentioned in another, separate, more advanced numbers exercise (my preference) or otherwise only be mentioned in the `after.md` document as I don't think it is information the student needs to know to be able to solve the exercise.

Reviewer 2: Sorry, some Stroop Effect happens for me, as I keep thinking these are about the concepts, rather than strictly what is needed to solve an exercise.

Because class was mentioned repeatedly, I thought inheritance was covered.

Reviewer: No worries.

Reviewer 3: ```suggestion
They are implemented through the `Integer` and `Float` classes.
```

> Both instances of `Float` and `Integer` implement methods to convert values from one to the other. As a number implemented through the `Integer` class has less precision than a number implemented through the `Float` class, converting from an instance of `Integer` to an instance of `Float` is safe because it does not result in any kind of data loss. However, converting from an instance of `Float` to an instance of `Integer` could mean losing data.

Reviewer 2: ```suggestion
The `Float` and `Integer` classes have methods that will coerce values from one to the other.  `Integer` numbers are precise to a whole unit, while `Float` has precision that is fractional to an whole number.  This means that coercing a float to an integer may result in loss of precision.
```
We can't "convert" a number from one type to another, as Ruby is strongly typed.  But Ruby does provide "coercion" methods that allow us to present one type as another.  An example of this are the methods `to_i`, `to_f`, and are similar to what we may have seen if we have already covered `Strings`, with `to_s`.
